probe-rs/src/architecture/arm/sequences/atsame5x.rs

```diff
@@ -0,0 +1,379 @@
+        while start.elapsed() < std::time::Duration::from_secs(8) {
+            let current_dsu_statusa = DsuStatusA::from(memory.read_word_8(DsuStatusA::ADDRESS)?);
+            if current_dsu_statusa.done() {
+                log::info!("Chip-Erase complete");
+                self.reset_hardware_with_extension(memory.get_arm_interface()?)?;
+
+                // We need to reconnect to target to finalize the unlock.
+                // Signal ReAttachRequired so that the session will try to re-connect
+                return Err(Error::Probe(DebugProbeError::ReAttachRequired));
+            } else if current_dsu_statusa.fail() {
+                return Err(Error::Other(anyhow::anyhow!("Chip-Erase Failed")));
+            }
+            std::thread::sleep(std::time::Duration::from_millis(250));
+        }
+
+        log::error!("Chip-Erase failed to complete within 8 seconds");
+        Err(Error::Probe(DebugProbeError::Timeout))
+    }
+
+    /// Perform a hardware reset in a way that puts the core into CPU Reset Extension
+    ///
+    /// CPU Reset Extension is a vendor specific feature that allows the CPU core to remain
+    /// in reset while the rest of the debugging subsystem can run and initialize itself.
+    ///
+    /// For more details see: 12.6.2 CPU Reset Extension in the SAM D5/E5 Family Data Sheet
+    ///
+    /// # Errors
+    /// Subject to probe communication errors
+    pub fn reset_hardware_with_extension(
+        &self,
+        interface: &mut dyn architecture::arm::communication_interface::SwdSequence,
+    ) -> Result<(), Error> {
+        let mut pins = architecture::arm::Pins(0);
+        pins.set_nreset(true);
+        pins.set_swdio_tms(true);
+        pins.set_swclk_tck(true);
+
+        let mut pin_values = architecture::arm::Pins(0);
+
+        // First set nReset, SWDIO and SWCLK to low.
+        let _ = interface.swj_pins(pin_values.0 as u32, pins.0 as u32, 0)?;
+        std::thread::sleep(std::time::Duration::from_millis(10));
+
+        // Next release nReset, but keep SWDIO and SWCLK low. This should put the device into
+        // reset extension ()
+        pin_values.set_nreset(true);
+        let _ = interface.swj_pins(pin_values.0 as u32, pins.0 as u32, 0)?;
+        std::thread::sleep(std::time::Duration::from_millis(20));
+
+        Ok(())
+    }
+
+    /// Release the CPU core from Reset Extension
+    ///
+    /// Clear the DSU Reset Extension bit, which releases the core from reset extension.
+    ///
+    /// # Errors
+    /// Subject to probe communication errors
+    pub fn release_reset_extension(&self, memory: &mut Memory) -> Result<(), Error> {
+        // clear the reset extension bit
+        let mut dsu_statusa = DsuStatusA(0);
+        dsu_statusa.set_crstext(true);
+        memory.write_8(DsuStatusA::ADDRESS, &[dsu_statusa.0])?;
+
+        let start = std::time::Instant::now();
+        while start.elapsed() < std::time::Duration::from_millis(100) {
```

Comment:
I am actually not sure how long this can take. The timing diagram in the data sheet suggests that its almost instantaneous (i.e. within one SWDCLK cycle from when the bit is written to). I chose 100ms out of the blue as something that should be more than enough to do the job but not long enough for the user notice any lag. 

Could you elaborate on what you mean with incremental, I'm not sure I understand?  Or you can point me to code that I should mimic?

Reply:
Looking at the code base, there's plenty of hardcoded timeouts. Especially when it comes to sequences. Others also seem to be hardcoded to be eg. 2 seconds.

Reply:
Yes, @Yatekii mentioned that as well in another thread. I think they are fine with it as is or?